test/e2e/network_segmentation.go

```diff
@@ -104,15 +105,15 @@ var _ = Describe("Network Segmentation", func() {
 					},
 					Entry("L2 primary UDN",
 						networkAttachmentConfigParams{
-							name:     nadName,
+							name:     randomNetworkMetaName(),
```

Comment:
this change is misleading.

In the NAD, the parameter that matters is actually the `networkName`, which is already auto-generated when omitted (it usually is):
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/test/e2e/multihoming_utils.go#L76

The name of the NAD is used to attach the workload to a network, whereas the name of the `network` - which is the ID of the network - is used to define the network. The latter is what must be unique across the namespaces, and it already is.

I think you generate the network names from the metadata name of the cUDN CR. You'll need randomize, sure. But not for NADs - that's already done.

Reply:
I had to make meta.name random because CUDN type is cluster scope, some tests fail every now and then on creating CUDN because there is already existing one with the same name.

The affected tests wrapped with `DescribeTableSubtree` to enable testing using CUDN type (similar to UDN type).
I had to change it for all cases (NAD and UDN types) due to the structure we have, its not optimal but I think its still better then duplicating test tables.

Make sense?

Reply:
No. Or I don't follow why you **have** to impact NADs - AFAIU, you don't. NADs are OK. NADs already generate their names. 

Your cUDN building function just needs to build UDNs from the generic data in which the network name is already generated - thus unique.

I think that's more aligned with the current status-quo.

Reply:
I had to impact NAD and UDN types due to the structure we have today in tests.
That is the fact we wrap all tests with `DescribeTableSubtree` to test the exact same scenarios using different CRD types 
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/e38f234624752f15755036cc2ca7975887786b51/test/e2e/network_segmentation.go#L1169-L1195

FWIU it works for NADs because there is some code that checks if no meta.name was specified it"ll generate one, its implicit and not trivial (at least for me).
I tried to avoid that for UDN and CUDN types to keep things simple.

Reply:
keeping things simple - to me - is to follow the established pattern, and adjust to the existing status-quo. Meaning, re-use the established pattern for NADs, and extend your code to generate a name based on the networkName parameter.

How is that non-trivial ? And how is that complex ? 
At most you can say it is not explicit - but I don't see why it has to be: it's an implementation detail. The important fact here is the network name is the ID of the network, and it must be generated. 

The fact the CRD you've implemented generates the `networkName` of the underlying NAD from the metadata.name of the CRD is IMHO an implementation detail.

Reply:
I got it wrong, it seem for NADs the networkName is generated if not specified I though it was the metadata.name, sorry for the noise.

Reply:
But you're affecting an object you have no need - at all - to affect. I still don't understand why you can't just carry the generated [attribute](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/9cf5a7707ca278b96fed631a8780fbf3ff7e32bb/test/e2e/multihoming_utils.go#L60) from the `networkAttachmentConfigParams` and use that as metadata.name of the CRD, in the CRD's constructor functions.

Reply:
I tried it, somthing similar where only CUDN cases are affected, generating random meta.name only at the test setup for CUDN cases [[1]](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/9cf5a7707ca278b96fed631a8780fbf3ff7e32bb/test/e2e/network_segmentation.go#L608-L621)
And having the `networkAttachmentConfigParams` cray that name into the test - [[2]](https://github.com/ormergi/ovn-kubernetes/commit/4e2ac804f0535c65cbe2c0f571e54bfae3ef88da#diff-2be1e01ba9cf5c059a4f4f0cd2b974861043381664e7102e3783f48842891d43R469-R477) (its a different branch)

It works, but require additional smelly change  in the "isolates overlapping CIDRs" tests table [[2]](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/9cf5a7707ca278b96fed631a8780fbf3ff7e32bb/test/e2e/network_segmentation.go#L433-L594) test table.

The "isolates overlapping CIDRs" works differently, it wont take meta.name like other tests (from the Entry part), 
it hard code names for object it creates; two namespaces, is each namespace, network object (NAD/UDN/CUDN) & pods.
The problem is that the objects names depends on each other, and the test assume objects has certain names.
For CUDN cases to work, its require to override the names used for the objects it creates (namepaces, pods and NAD/UDN/CUDN) [[1]](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/9cf5a7707ca278b96fed631a8780fbf3ff7e32bb/test/e2e/network_segmentation.go#L608-L621).

IMO the current solution is easier to understand thus I prefer it over the above alternative, that is due to how the  "isolates overlapping CIDRs" structured today.
WDYT?

Refactoring the "isolates overlapping CIDRs" requires some additional work and time I prefer to keep outside this PR socpe.

Reply:
I essentially think the tests - and the helpers, as they exist today - are meant to be used in a certain way.

The fact we've chosen in the past to embed everything in tables (especially once we've nested the NAD/CRD table in ...) only made this contract more important: *everything * should adhere to these principles, otherwise, other areas in the tests will be impacted.

Sadly, part of the tests is not using the generated names. I understand that. Probably there's a reason, maybe there isn't one - but, the ugly truth is it has impacts elsewhere. **Everywhere** actually because we have nested the tables. Which, was a good choice. 

As you said, the proper solution is refactoring the "isolates overlapping CIDRs" tests, breaking the dependency it has on object's `metadata.name`; I'm aware that takes more work.

What you propose here is an immediate patch that works, but diverges us further from the was the helpers are supposed to be used. It's something that will impact every test, when it shouldn't, and it definitely feels you're hitting the code with an hammer to make it bend to your use case. I'm afraid in the long term, this will hurt us more.

I'd still vote for refactoring that test, but I won't block this PR over this.

Reply:
Thanks for the feedback.

> Sadly, part of the tests is not using the generated names. I understand that. Probably there's a reason, maybe there isn't one - but, the ugly truth is it has impacts elsewhere. Everywhere actually because we have nested the tables. Which, was a good choice.

All tests use hard-coded metadata.names, it works because each test run in its own namespace, no need to generate names.
Including cases that cover NAD usage, the do generate the network-name, if not given.
On CUDN usage cases it wont go for reasons mentioned above (its cluster scope object, must have unique name).

> What you propose here is an immediate patch that works, but diverges us further from the was the helpers are supposed to be used. It's something that will impact every test, when it shouldn't, and it definitely feels you're hitting the code with an hammer to make it bend to your use case. I'm afraid in the long term, this will hurt us more.

If you think the other alternative you suggest is better - we can go with that (patching the used names inside "isolates overlapping CIDRs"), eventually it doesnt impact the entire nested table.
Otherwise, I will appreciate your approval on this PR :)

Reply:
I don't know how else to indicate that I think that's better. I've been arguing for that since the beginning. 

I will not approve, because I think this is wrong. I have explained what I think is correct and why in the previous comments. 

I also don't plan to block the PR; while I think this is wrong, if you can find someone to review and approve as is, be my guest.

Reply:
If I am following this discussion correctly, I think all we need to do is keep NAD name as it was, but change https://github.com/ovn-kubernetes/ovn-kubernetes/blob/be2dfda65ad8ecec188f842233108768e1ab4085/test/e2e/network_segmentation.go#L1407-L1413 to generate unique CUDN.name instead of just using `params.name` 
@ormergi will that solve the problem?

Reply:
Unfortunately it wont solve it for all tests, the "isolates overlapping CIDRs" cases using CUDN will break, 
because there are dependencies on the CUDN object (e.g.: test expects CUDN name to be A, but the name is now B). 

I will change it to what suggested above: generate random meta.name for CUDNs here [[1]](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/6b8ef77825556517e2116d9ac3e756867f815a48/test/e2e/network_segmentation.go#L608), and have the `networkAttachmentConfigParams` cray that name into the tests - [[2]](https://github.com/ormergi/ovn-kubernetes/commit/4e2ac804f0535c65cbe2c0f571e54bfae3ef88da#diff-2be1e01ba9cf5c059a4f4f0cd2b974861043381664e7102e3783f48842891d43R613-R615) (its a different branch)

Reply:
Done